internal/common/persistence/serializer.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"encoding/json"
+	"fmt"
+	s "go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/internal/common"
+)
+
+type (
+	// PayloadSerializer is used by persistence to serialize/deserialize history event(s) and others
+	// It will only be used inside persistence, so that serialize/deserialize is transparent for application
+	PayloadSerializer interface {
+		SerializeBatchEvents(batch []*s.HistoryEvent, encodingType s.EncodingType) (*s.DataBlob, error)
+		// deserialize history events
+		DeserializeBatchEvents(data *s.DataBlob) ([]*s.HistoryEvent, error)
+	}
+
+	// CadenceSerializationError is an error type for cadence serialization
+	CadenceSerializationError struct {
+		msg string
+	}
+
+	// CadenceDeserializationError is an error type for cadence deserialization
+	CadenceDeserializationError struct {
+		msg string
+	}
+
+	// UnknownEncodingTypeError is an error type for unknown or unsupported encoding type
+	UnknownEncodingTypeError struct {
+		encodingType s.EncodingType
+	}
```

Comment:
I think I may lack some context here. But is there any special reason for creating separate struct for those errors, instead of using `errors.New()` or `fmt.Errorf()`?